src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,344 @@
+  private static final long   ACCELERATE_CHUNK_INTERVAL = 500L;
+  private static final double JITTER_MAX                = 0.20;
+
+  private final String              serverApiUrl;
+  private final ReplicatedJedisPool jedisPool;
+  private final AccountsManager     accountsManager;
+  private final Client              client;
+
+  private final String       workerId;
+  private final SecureRandom random;
+
+  private final UnlockOperation unlockOperation;
+
+  private boolean running;
+  private boolean finished;
+
+
+  public DirectoryReconciler(ContactDiscoveryConfiguration cdsConfig,
+                             ReplicatedJedisPool jedisPool,
+                             AccountsManager accountsManager)
+      throws IOException
+  {
+    DirectoryConfiguration directoryConfig = cdsConfig.getDirectoryConfiguration();
+    this.serverApiUrl    = directoryConfig.getServerApiUrl();
+    this.jedisPool       = jedisPool;
+    this.accountsManager = accountsManager;
+
+    this.random = new SecureRandom();
+
+    byte[] workerIdBytes = new byte[16];
+    random.nextBytes(workerIdBytes);
+    this.workerId = Hex.toString(workerIdBytes);
+
+    this.unlockOperation = new UnlockOperation(jedisPool);
+
+    SslConfigurator sslConfig = SslConfigurator.newInstance()
+                                               .securityProtocol("TLSv1.2");
+    try {
+      sslConfig.trustStore(initializeKeyStore(directoryConfig.getServerApiCaCertificate()));
+    } catch (CertificateException ex) {
+      logger.error("error reading serverApiCaCertificate from contactDiscovery config", ex);
+    }
+
+    this.client = ClientBuilder.newBuilder()
+                               .register(HttpAuthenticationFeature.basic("signal", directoryConfig.getServerApiToken().getBytes()))
+                               .sslContext(sslConfig.createSSLContext())
+                               .build();
+  }
+
+  private static KeyStore initializeKeyStore(String pemCaCertificate)
+          throws CertificateException
+  {
+    try {
+      PEMReader       reader      = new PEMReader(new InputStreamReader(new ByteArrayInputStream(pemCaCertificate.getBytes())));
+      X509Certificate certificate = (X509Certificate) reader.readObject();
+
+      if (certificate == null) {
+        throw new CertificateException("No certificate found in parsing!");
+      }
+
+      KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      keyStore.load(null);
+      keyStore.setCertificateEntry("ca", certificate);
+      return keyStore;
+    } catch (IOException | KeyStoreException ex) {
+      throw new CertificateException(ex.toString());
+    } catch (NoSuchAlgorithmException ex) {
+      throw new AssertionError(ex);
+    }
+  }
+
+  @Override
+  public synchronized void start() throws Exception {
+    running  = true;
+    finished = false;
+    new Thread(this).start();
+  }
+
+  @Override
+  public synchronized void stop() throws Exception {
```

Comment:
Doesn't this need to notifyAll() if you're counting on sleepWhileRunning to wake up?

Reply:
yes, got lost in refactor. will fix.